### PR TITLE
Cyble Events Integration

### PR DESCRIPTION
**Pull request by cyble-dev**
Integration pack for Cyble Event incidents

## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
Cyble Events an incident integration pack 

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [x] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by GuyAfik** [comment]
Hi @cyble-dev Thank you for your incredible work in the contribution! Added some comments, if you need any help or any questions reach me out in slack

**Comment by tomer-pan** [comment]
@cyble-dev Please note that all of the json files of the incident fields should not include the `incidentFields` key. Instead, each incident field should only include an incident field object. Please see this file as an example for the correct json structure:
https://github.com/demisto/content/blob/master/Packs/Phishing/IncidentFields/incidentfield-Attachment_Count.json

Also, please note that there are 3 common incident fields which you can use instead of creating custom ones:
- Event ID
- Event Type
- Event Names

File: Packs/CybleEvents/Integrations/CybleEvents/README.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+| **Parameter** | **Description** | **Required** | 
+| --- | --- | --- |
+| from | Returns records started with given value | False | 
+| limit | Number of records to return (max 1000). Using a smaller limit will get faster responses. | False | 
+| start_date | Timeline start date in the format "YYYY-MM-DD". Need to used with end_date as timeline range. | False | 
+| end_date | Timeline end date in the format "YYYY-MM-DD". Need to used with start_date as timeline range. | False | 
+| type | Returns record by type like (CIDR, CVE, domain, email, FileHash-IMPHASH, FileHash-MD5, FileHash-PEHASH,
+FileHash-SHA1, FileHash-SHA256, FilePath, hostname, IPv4, IPv6, Mutex, NIDS, URI, URL, YARA, osquery, Ja3,
+Bitcoinaddress, Sslcertfingerprint). | False | 
+| keyword | Returns records for the specified keyword | False|
```
Comment (GuyAfik) [inline Packs/CybleEvents/Integrations/CybleEvents/README.md:27]: The README structure of the commands should be as the following (this is just an example):
```
### pan-os-delete-address
***
Delete an address object


#### Base Command

`pan-os-delete-address`
#### Input

| **Argument Name** | **Description** | **Required** |
| --- | --- | --- |
| name | Name of the address to delete. | Required | 
| device-group | The device group for which to return addresses (Panorama instances). | Optional | 


#### Context Output

| **Path** | **Type** | **Description** |
| --- | --- | --- |
| Panorama.Addresses.Name | string | Address name that was deleted. | 
| Panorama.Addresses.DeviceGroup | String | Device group for the address \(Panorama instances\). | 


#### Command Example
```!pan-os-delete-address name="address_test_pb"```

#### Context Example
```json
{
    "Panorama": {
        "Addresses": {
            "Name": "address_test_pb"
        }
    }
}
```

#### Human Readable Output

>Address was deleted successfully.

It would be nice if you can change to that format. To automatically generate that you can use the SDK generate-docs command, see how to use [here](https://github.com/demisto/demisto-sdk/blob/master/demisto_sdk/commands/generate_docs/README.md)

File: Packs/CybleEvents/Integrations/CybleEvents/CybleEvents_test.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+        content = json.load(f)
+    return content
+
+
+def test_module():
+    """
+    Test the basic test command for Cyble Events
+    :return:
+    """
+    pass
```
Comment (GuyAfik) [inline Packs/CybleEvents/Integrations/CybleEvents/CybleEvents_test.py:22]: why is that necessary?

File: Packs/CybleEvents/Integrations/CybleEvents/CybleEvents_description.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+| start_date | Timeline start date in the format "YYYY/MM/DD" | True |
+| end_date | Timeline end date in the format "YYYY/MM/DD" | True |
+| order_by | Sorting order for alert fetch either Ascending or Descending | True |
+
+
+**!cyble-vision-fetch-event-detail**
+| **Parameter** | **Description** | **Required** |
+| --- | --- | --- |
+| event_type | Event Type of the Incident | True |
+| event_id | Event ID of the incident | True |
```
Comment (GuyAfik) [inline Packs/CybleEvents/Integrations/CybleEvents/CybleEvents_description.md:42]: not required in a description file, you can remove it. for that we have the integration README

File: Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.yml [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+  | Returns records for the specified keyword | False|\n\n\n**!cyble-vision-fetch-events**\n|
+  **Parameter** | **Description** | **Required** |\n| --- | --- | --- |\n| from |
+  Returns records for the timeline starting from given indice | True |\n| limit |
+  Number of records to return (max 50). Using a smaller limit will get faster responses.
+  | True|\n| start_date | Timeline start date in the format \"YYYY/MM/DD\" | True
+  |\n| end_date | Timeline end date in the format \"YYYY/MM/DD\" | True |\n| order_by
+  | Sorting order for alert fetch either Ascending or Descending | True |\n\n\n**!cyble-vision-fetch-event-detail**\n|
+  **Parameter** | **Description** | **Required** |\n| --- | --- | --- |\n| event_type
+  | Event Type of the Incident | True |\n| event_id | Event ID of the incident | True
+  |"
```
Comment (GuyAfik) [inline Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.yml:39]: I think you can remove it

File: Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py [hunk 5]
```diff
@@ -0,0 +1,493 @@
...
+        response = requests.request('{}'.format(str(method).upper()), url, headers=headers, data=payload)
+
+        resp = {}
+        try:
+            if response.status_code == 200:
+                resp = response.json()
+        except Exception as e:
+            demisto.error('Exception while fetching the event details {}'.format(e))
+
+        if response.status_code == 200 and 'success' in resp.keys() and (resp.get('success') or False):
```
Comment (GuyAfik) [inline Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py:169]:
```suggestion
        if response.status_code == 200 and (resp.get('success') or False):
```

File: Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py [hunk 6]
```diff
@@ -0,0 +1,493 @@
...
+            events_data.extend(resp.get('events', []))
+            if resp['total_count'] != len(events_data) and len(events_data) <= MAX_EVENT_ITEMS:
```
Comment (GuyAfik) [inline Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py:171]:
```suggestion
            if resp.get('total_count') != len(events_data) and len(events_data) <= MAX_EVENT_ITEMS:
```

File: Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py [hunk 7]
```diff
@@ -0,0 +1,493 @@
...
+    if 'total_alert_count' not in last_run.keys():
+        last_run['total_alert_count'] = 0
+    if 'fetched_alert_count' not in last_run.keys():
+        last_run['fetched_alert_count'] = 0
+    if 'event_pull_start_date' not in last_run.keys():
+        last_run['event_pull_start_date'] = date.today().strftime("%Y/%m/%d")
+
+    params = {
+        'token': token,
+        'from': int(last_run.get('fetched_alert_count', '0')),
```
Comment (GuyAfik) [inline Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py:379]:
```suggestion
        'from': arg_to_number(last_run.get('fetched_alert_count', '0')),
```

File: Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py [hunk 8]
```diff
@@ -0,0 +1,493 @@
...
+
+            return_results(cyble_fetch_iocs(client, 'POST', args))
+
+        elif demisto.command() == 'cyble-vision-fetch-events':
+            # This is the call made when cyble-fetch-events command.
+            args['order_by'] = str(args['order_by']).title()
+            if 'start_date' not in args.keys():
+                args['start_date'] = datetime.today().strftime('%Y/%m/%d')
+            if 'end_date' not in args.keys():
+                args['end_date'] = datetime.today().strftime('%Y/%m/%d')
```
Comment (GuyAfik) [inline Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py:479]:
```suggestion
            if not args.get('start_date'):
                args['start_date'] = datetime.today().strftime('%Y/%m/%d')
            if not args.get('end_date'):
                args['end_date'] = datetime.today().strftime('%Y/%m/%d')
```

File: Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py [hunk 9]
```diff
@@ -0,0 +1,493 @@
...
+        headers = {
+            'X-API-KEY': '{}'.format(params.get('token',''))
+        }
+        url = urljoin(self._base_url, etypeurl)
+
+        response = requests.request(method, url, headers=headers, data=payload)
+        try:
+            resp = response.json()
+
+            if 'success' in resp.keys() and resp['success'] is True:
```
Comment (GuyAfik) [inline Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py:55]:
```suggestion
            if resp.get('success') or False:
```

File: Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py [hunk 10]
```diff
@@ -0,0 +1,493 @@
...
+        :param iocurl: API URL Suffix to be used
+        :param params: parameters to be used as part of request
+        :return:indicator details as JSON
+        """
+        ioc_data = {}
+        token = params.get('token', '')
+        payload = {
+            'token': '{}'.format(token),
+            'from': int(params.get('from', '')),
+            'limit': int(params.get('limit', '50')),
```
Comment (GuyAfik) [inline Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py:77]:
```suggestion
            'from': arg_to_number(params.get('from', '0')),
            'limit': arg_to_number(params.get('limit', '50')),
```

File: Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py [hunk 11]
```diff
@@ -0,0 +1,493 @@
...
+
+        try:
+            resp = response.json()
+            if 'count' in resp.keys():
+                ioc_data = resp
+            else:
+                ioc_data = {"error": "Failed to Fetch Taxiis !!"}
+                demisto.error("Error trying to Fetch IOC's {}".format(resp))
+        except Exception as e:
+            demisto.error("Error: [{}] for response [{}]".format(e, resp))
```
Comment (GuyAfik) [inline Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py:99]:
```suggestion
        except Exception as e:
            demisto.error("Error: [{}] for response [{}]".format(e, resp))
            raise e
```

File: Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py [hunk 12]
```diff
@@ -0,0 +1,493 @@
...
+        response = requests.request(method, url, headers=headers, data=payload)
+        try:
+            resp = response.json()
+
+            if 'success' in resp.keys() and resp['success'] is True:
+                eventTypeAlias = resp['data']
+            else:
+                demisto.error("Error trying to Fetch EventTypess {}".format(resp))
+        except Exception as e:
+            demisto.error("Exception with Fetch EventTypes [{}]".format(e))
```
Comment (GuyAfik) [inline Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py:60]:
```suggestion
            demisto.error("Exception with Fetch EventTypes [{}]".format(e))
            raise e
```

File: Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py [hunk 13]
```diff
@@ -0,0 +1,493 @@
...
+        headers = {
+            "Cookie": "XSRF-TOKEN={}".format(token)
+        }
+        url = urljoin(self._base_url, iocurl)
+
+        response = requests.request('{}'.format(str(method).upper()), url, headers=headers, data=payload, files=files)
+
+        try:
+            resp = response.json()
+            if 'count' in resp.keys():
```
Comment (GuyAfik) [inline Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py:93]:
```suggestion
            if resp.get('count'):
```

File: Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py [hunk 14]
```diff
@@ -0,0 +1,493 @@
...
+
+        try:
+            resp = response.json()
+
+            if 'success' in resp.keys() and (resp.get('success') or False):
+                events_data = resp.get('data', {}).get('results')
+            else:
+                demisto.error("Error trying to Fetch Events {}".format(resp))
+        except Exception as e:
+            demisto.error("Exception with Fetch Events [{}]".format(e))
```
Comment (GuyAfik) [inline Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py:136]:
```suggestion
            demisto.error("Exception with Fetch Events [{}]".format(e))
            raise e
```

File: Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py [hunk 15]
```diff
@@ -0,0 +1,493 @@
...
+                maxResults=max_results
+            )
+            demisto.incidents(incidents)
+
+        elif demisto.command() == 'cyble-vision-fetch-iocs':
+            # This is the call made when cyble-fetch-iocs command.
+            if 'start_date' not in args.keys():
+                args['start_date'] = datetime.today().strftime('%Y-%m-%d')
+            if 'end_date' not in args.keys():
+                args['end_date'] = datetime.today().strftime('%Y-%m-%d')
```
Comment (GuyAfik) [inline Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py:469]:
```suggestion
            if not args.get('start_date):
                args['start_date'] = datetime.today().strftime('%Y-%m-%d')
            if not args.get('end_date'):
                args['end_date'] = datetime.today().strftime('%Y-%m-%d')
```

**Review by GuyAfik — COMMENTED** [review]
Hi, very nice work on fixing those comments! a few points:
1) See [here](https://app.circleci.com/pipelines/github/demisto/content/154136/workflows/cf3e8daa-846e-4c7c-af20-c522a57bb89f/jobs/500129) that you have failing tests
2) See [here](https://app.circleci.com/pipelines/github/demisto/content/154128/workflows/aaaa59ff-d05a-40ca-9190-990868ca3119/jobs/500105) the build errors, if you are not able to resolve those, let me know and ill help you

**Comment by cyble-dev** [comment]
> @cyble-dev Please note that all of the json files of the incident fields should not include the `incidentFields` key. Instead, each incident field should only include an incident field object. Please see this file as an example for the correct json structure: https://github.com/demisto/content/blob/master/Packs/Phishing/IncidentFields/incidentfield-Attachment_Count.json
> 
> Also, please note that there are 3 common incident fields which you can use instead of creating custom ones:
> 
> * Event ID
> * Event Type
> * Event Names

have fixed the alignments, and will go with custom names as it would be easy to manage as one set

**Comment by cyble-dev** [comment]
all the reviews and validation suggestions are added

File: Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py [hunk 16]
```diff
@@ -0,0 +1,500 @@
...
+        :param method: requests method to perform the desired action to be performed
+        :param etypeurl: event type URL path
+        :param params: parameters to be used as part of request
+        :return: event types along with alias as a JSON
+        """
+        eventTypeAlias = None
+
+        payload = {}
+        headers = {
+            'X-API-KEY': '{}'.format(params['token'])
```
[the remark this answers is not in the document]
Comment (cyble-dev) [reply Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py]: changes are added

File: Packs/CybleEvents/Integrations/CybleEvents/README.md (hunk 1 above)
Comment (cyble-dev) [reply Packs/CybleEvents/Integrations/CybleEvents/README.md]: made the changes

File: Packs/CybleEvents/Integrations/CybleEvents/CybleEvents_description.md (hunk 3 above)
Comment (cyble-dev) [reply Packs/CybleEvents/Integrations/CybleEvents/CybleEvents_description.md]: updated the changes

File: Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.yml (hunk 4 above)
Comment (cyble-dev) [reply Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.yml]: done with changes

File: Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py [hunk 17]
```diff
@@ -0,0 +1,497 @@
...
+            'Content-Type': 'application/json'
+        }
+
+        url = urljoin(self._base_url, eventurl)
+        response = requests.request('{}'.format(str(method).upper()), url, headers=headers, data=payload)
+
+        try:
+            resp = response.json()
+
+            if 'success' in resp.keys() and (resp.get('success') or False):
```
Comment (GuyAfik) [inline Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py:133]:
```suggestion
            if (resp.get('success') or False):
```

Comment (cyble-dev) [reply Packs/CybleEvents/Integrations/CybleEvents/CybleEvents.py]: updated

**Comment by GuyAfik** [comment]
@tomer-pan can you please re-review? changes were all done since the demo.

File: Packs/CybleEvents/IncidentFields/incident_cybleeventstype.json [hunk 18]
```diff
@@ -0,0 +1,30 @@
+{
```
Comment (tomer-pan) [inline Packs/CybleEvents/IncidentFields/incident_cybleeventstype.json:1]: @cyble-dev Please use the common field "Event Type" instead.

File: Packs/CybleEvents/IncidentFields/incident_cybleeventsname.json [hunk 19]
```diff
@@ -0,0 +1,30 @@
+{
```
Comment (tomer-pan) [inline Packs/CybleEvents/IncidentFields/incident_cybleeventsname.json:1]: @cyble-dev Please use the common field "Event Names" instead.

File: Packs/CybleEvents/IncidentFields/incident_cybleeventsid.json [hunk 20]
```diff
@@ -0,0 +1,30 @@
+{
```
Comment (tomer-pan) [inline Packs/CybleEvents/IncidentFields/incident_cybleeventsid.json:1]: @cyble-dev Please use the common field "Event ID" instead.

File: Packs/CybleEvents/IncidentFields/incident_cybleeventstype.json (hunk 18 above)
Comment (cyble-dev) [reply Packs/CybleEvents/IncidentFields/incident_cybleeventstype.json:1]: will go with the custom name as it would be easy to manage as one entity and tightly coupled

File: Packs/CybleEvents/IncidentFields/incident_cybleeventsid.json (hunk 20 above)
Comment (cyble-dev) [reply Packs/CybleEvents/IncidentFields/incident_cybleeventsid.json:1]: will go with the custom name as it would be easy to manage as one entity and tightly coupled

File: Packs/CybleEvents/IncidentFields/incident_cybleeventstype.json (hunk 18 above)
Comment (tomer-pan) [reply Packs/CybleEvents/IncidentFields/incident_cybleeventstype.json:1]: @cyble-dev We are trying to minimize custom fields due to an overload we currently have.
The guidelines for incident fields is to use the common fields when it's possible and a lot of them are associated to different incident types (you can just associate them to your incident type and use it). Same for the other 2 fields I mentioned. If it's not possible for you to use the common fields for technical reasons, then you can create custom ones, but if the meaning and the usage of the custom field is the same as the common one, you should use the common field.

Comment (cyble-dev) [reply Packs/CybleEvents/IncidentFields/incident_cybleeventstype.json:1]: yes the field type is different for one. Also had faced issue in fetching the assigned data from these fields into layout. With this technical hindrance faced while implementation, going for these custom fields.

File: Packs/CybleEvents/IncidentFields/incident_cybleeventsid.json (hunk 20 above)
Comment (cyble-dev) [reply Packs/CybleEvents/IncidentFields/incident_cybleeventsid.json]: used common field here

File: Packs/CybleEvents/IncidentFields/incident_cybleeventsname.json (hunk 19 above)
Comment (cyble-dev) [reply Packs/CybleEvents/IncidentFields/incident_cybleeventsname.json:1]: the type of common field is different so will be using custom field here

File: Packs/CybleEvents/IncidentFields/incident_cybleeventstype.json (hunk 18 above)
Comment (cyble-dev) [reply Packs/CybleEvents/IncidentFields/incident_cybleeventstype.json]: used common field for this

File: Packs/CybleEvents/Classifiers/classifier-Cyble_Intel_Alert_Mapper.json [hunk 21]
```diff
@@ -15,11 +15,11 @@
                     },
                     "simple": ""
                 },
-                "Cyble Events ID": {
+                "Event ID": {
                     "complex": {
                         "accessor": "",
                         "filters": [],
-                        "root": "cybleeventsid",
+                        "root": "eventid",
```
Comment (tomer-pan) [inline Packs/CybleEvents/Classifiers/classifier-Cyble_Intel_Alert_Mapper.json:22]: @cyble-dev Why did you change the root? When mapping incident fields, you should map the relevant incident field (like "Event ID") to the relevant key in the incident json. So for example if the original json key is named "cybleeventid" for example, the root should contain this value. Same for all of the other keys.

**Review by GuyAfik — APPROVED** [review]
Looks good! @cyble-dev thanks for your amazing work!
